### ISSUE TITLE
Use Zopfli DeflateEncoder with customizable buffer size

### ISSUE THREAD

**Issue by Pr0methean** [shssoichiro/oxipng#521]
I have a program that creates large PNGs (up to 48 MiB of raw pixel data, but usually 12 or 16 MiB) and uses Zopfli as the deflater for PNG encoding. At these sizes, compression ratios are usually between 99.4% and 99.9%. Compression is currently slow because of the splits at every ZOPFLI_MASTER_BLOCK_SIZE (currently 1 million) bytes. When I raised this issue at https://github.com/zopfli-rs/zopfli/issues/18, they suggested to instead use a DeflateEncoder wrapping a large BufWriter. Could we please have the option to do that? The best way would be to have the user supply a mutably borrowed DeflateEncoder, so that we could reuse the buffer for multiple images.

**Comment by AlexTMjugador** [comment]
Putting on my OxiPNG contributor hat, I think this might be part of the explanation for #414. I'm busy with other things at the moment, but I feel it's worth taking a look at how performance compares with different master block sizes here.

**Comment by Pr0methean** [comment]
I'll experiment with this in a fork.

**Comment by Pr0methean** [comment]
See https://github.com/Pr0methean/oxipng/commit/97248f799f3895e808a9720e296b2717927a468a.

**Comment by TPS** [comment]
> I'll experiment with this in a fork.

@Pr0methean How are the results from this fork/commit?

**Comment by andrews05** [comment]
@AlexTMjugador Afaik, zopflipng doesn't alter the ZOPFLI_MASTER_BLOCK_SIZE - it's a compile time constant which defaults to 1MB. So surely that wouldn't explain the other issue?

**Comment by AlexTMjugador** [comment]
> So surely that wouldn't explain the other issue?

That constant is indeed unchanged, but it might be the case (I'm not sure right now, I may be wrong) that ZopfliPNG exercises a different code path that does not use that constant for segmenting the input data. After all, practical PNG images are not _big_, and that master block size was mostly meant to deal with big datasets such as the enwiki corpus.

**Comment by Pr0methean** [comment]
@TPS So far, the speed seems to be about the same, but it has the advantage that Zopfli can choose strategically where to split the file.

**Comment by TPS** [comment]
@Pr0methean That seems encouraging.  Why not make this an official PR here (draft, if you're not satisfied &/or done w/ tweaking it), referring this & the above issue, so, whenever the rest are merged by @shssoichiro in preparation for the next version & binary update, your work can benefit all of us? 🙇🏾‍♂️

**Comment by TPS** [comment]
> - As noted in the comments, part of the code is "forked" from zopfli-rs. I'm not sure what's different, but surely it would be better to get the required changes submitted to the zopfli library? It doesn't seem like a great idea to keep partially forked code here.

@Pr0methean Per andrews05's comment, did you happen to PR to `zopli-rs`?  I'm curious to find out how it'd do.

**Comment by Pr0methean** [comment]
Yes; my PR was remade as https://github.com/shssoichiro/oxipng/pull/530 and merged.

**Comment by TPS** [comment]
> Yes; my PR was remade as https://github.com/shssoichiro/oxipng/pull/530 and merged.

& then [reverted](https://github.com/shssoichiro/oxipng/commit/31e796c4e878763794529bd8f071ab743c2ccf77)
, favoring [PR'ing @ `zopli-rs`](https://github.com/shssoichiro/oxipng/pull/530#issuecomment-1627557583).  I was hoping you'd done so.

**Comment by TPS** [comment]
@Pr0methean Since this is merged is `zopfli-rs` @ https://github.com/zopfli-rs/zopfli/commit/21d552107a003ac312c49f7346bba3f7722458a8, close FTW?
